In [13]:
import pandas as pd
import numpy as np

#### Pandas allows chaining together functions that expect Series, DataFrames or GroupBy objects

#### create dataframe by reading the data set

In [84]:
df = pd.read_csv("titanic.csv", sep="\t", index_col='PassengerId')
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

#### starts the index from 0

In [17]:
def transform_index(df):
    df.index = np.arange(df.shape[0])
    return df

#### we noticed few NaNs in "Cabin" column

In [37]:
df["Cabin"].isna().sum()

125

#### Actually we have lots of (=125) NaNs
#### We would like to replace NaN by looking at the "Sex" column, It is replaced by the cabin which was frequent for a particular sex in that row

In [69]:
def replace_nans_cabin(df):
    most_common = df.groupby('Sex')['Cabin'].apply(lambda x: x.value_counts().index[0]).reset_index()
    df.loc[(df['Cabin'].isnull()) & (df.Sex == 'female'), 'Cabin'] = most_common[most_common['Sex']=='female']['Cabin'][0]
    df.loc[(df['Cabin'].isnull()) & (df.Sex == 'male'), 'Cabin'] = most_common[most_common['Sex']=='male']['Cabin'][1]
    return df

#### We also replace NaNs in "Age" column by replacing it with the mean age of that particular sex

In [89]:
def replace_nans_age(df):
    df['Age'] = df["Age"].fillna(df.groupby("Sex")["Age"].transform('mean'))
    return df

#### Creates pipeline

In [90]:
(df.pipe(transform_index)
.pipe(replace_nans_cabin)
 .pipe(replace_nans_age)
)

Survived  Pclass                                               Name  \
0           0       3                            Braund, Mr. Owen Harris   
1           1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2           1       3                             Heikkinen, Miss. Laina   
3           1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4           0       3                           Allen, Mr. William Henry   
5           0       3                                   Moran, Mr. James   
6           0       1                            McCarthy, Mr. Timothy J   
7           0       3                     Palsson, Master. Gosta Leonard   
8           1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9           1       2                Nasser, Mrs. Nicholas (Adele Achem)   
10          1       3                    Sandstrom, Miss. Marguerite Rut   
11          1       1                           Bonnell, Miss. Elizabeth   
12          0       3                     Saundercock, Mr. William Henry   
13          0       3                        Andersson, Mr. Anders Johan   
14          0       3               Vestrom, Miss. Hulda Amanda Adolfina   
15          1       2                   Hewlett, Mrs. (Mary D Kingcome)    
16          0       3                               Rice, Master. Eugene   
17          1       2                       Williams, Mr. Charles Eugene   
18          0       3  Vander Planke, Mrs. Julius (Emelia Maria Vande...   
19          1       3                            Masselmani, Mrs. Fatima   
20          0       2                               Fynney, Mr. Joseph J   
21          1       2                              Beesley, Mr. Lawrence   
22          1       3                        McGowan, Miss. Anna "Annie"   
23          1       1                       Sloper, Mr. William Thompson   
24          0       3                      Palsson, Miss. Torborg Danira   
25          1       3  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...   
26          0       3                            Emir, Mr. Farred Chehab   
27          0       1                     Fortune, Mr. Charles Alexander   
28          1       3                      O'Dwyer, Miss. Ellen "Nellie"   
29          0       3                                Todoroff, Mr. Lalio   
..        ...     ...                                                ...   
126         0       3                                McMahon, Mr. Martin   
127         1       3                          Madsen, Mr. Fridtjof Arne   
128         1       3                                  Peter, Miss. Anna   
129         0       3                                 Ekstrom, Mr. Johan   
130         0       3                               Drazenoic, Mr. Jozef   
131         0       3                     Coelho, Mr. Domingos Fernandeo   
132         0       3     Robins, Mrs. Alexander A (Grace Charity Laury)   
133         1       2      Weisz, Mrs. Leopold (Mathilde Francoise Pede)   
134         0       2                     Sobey, Mr. Samuel James Hayden   
135         0       2                                 Richard, Mr. Emile   
136         1       1                       Newsom, Miss. Helen Monypeny   
137         0       1                        Futrelle, Mr. Jacques Heath   
138         0       3                                Osen, Mr. Olaf Elon   
139         0       1                                 Giglio, Mr. Victor   
140         0       3                      Boulos, Mrs. Joseph (Sultana)   
141         1       3                           Nysten, Miss. Anna Sofia   
142         1       3  Hakkarainen, Mrs. Pekka Pietari (Elin Matilda ...   
143         0       3                                Burke, Mr. Jeremiah   
144         0       2                         Andrew, Mr. Edgardo Samuel   
145         0       2                       Nicholls, Mr. Joseph Charles   
146         1       3       Andersson, Mr. August Edv